Add cursor "smoke" trails

goodtrailer opened the pull request:
Addresses #5721.

- [x] Depends on ppy/osu-resources#214

`LegacySmoke` implementation copies stable closely (#20372). Gives `DefaultSmoke` temp anims.

Uses `DrawNode`s because a `Sprite`-based implementation I tried performed poorly, though I'm not sure if that's due to having so many `Sprite`s (along with their `Transform`s) is laggy or if the implementation was just bad. `Sprite`s are a lot easier to work with though, so a well performing implementation through `Sprite`s would probably be better. And I think a shader (like with cursor trails, as eluded to in the above discussion thread) would perform better. So this is admittedly a bit of a weird middle ground.

Notes/caveats:
* `LegacySmoke` point spacing ("accepted width") calculation is different from stable
* `LegacySmoke` scale animation is different from stable, because as far as I could tell, the scale anim in stable's code doesn't actually do anything (so I modified it to look like how I think it was probably meant to look like)
* `LegacySmoke` time calculations are pretty intricate (specifically L109-L111)
* `Smoke` has a max duration of 60s (not sure if this limit is really necessary)

goodtrailer committed:
Basic smoke path implementation

goodtrailer committed:
Full legacy smoke implementation and temp default smoke

goodtrailer committed:
Show smoke in replays

goodtrailer committed:
Fix InspectCode errors

goodtrailer committed:
Remove smoke from key overlay

goodtrailer committed:
Copy stable smoke's fade/alpha values, blending, scale, and rotation

goodtrailer committed:
Remove hacky LifetimeEnd workaround

goodtrailer committed:
Temp default smoke scale/rotation anims

goodtrailer committed:
Scale smoke radius based on texture width

Feodor0090 commented on osu.Game.Rulesets.Osu/OsuInputManager.cs:
```diff
@@ -56,6 +57,20 @@ protected override bool HandleMouseTouchStateChange(TouchStateChangeEvent e)
             return base.HandleMouseTouchStateChange(e);
         }
 
+        public override void Attach(KeyCounterDisplay keyCounter)
+        {
+            var receptor = new ActionReceptor(keyCounter);
+
+            KeyBindingContainer.Add(receptor);
+
+            keyCounter.SetReceptor(receptor);
+            keyCounter.AddRange(new[]
+            {
+                new KeyCounterAction<OsuAction>(OsuAction.LeftButton),
+                new KeyCounterAction<OsuAction>(OsuAction.RightButton),
+            });
+        }
+
```
While playtesting this, i even thought that the lack of smoke button in counter is a bug. IMO it's better to show it there.

ItsShamed replied:
Yeah, like other custom rulesets do it as well it's not a big deal to show it.

peppy replied:
I agree with showing it.

goodtrailer replied:
Hmm ok, though some tests will have to be rewritten (and I'm not really familiar with writing tests)

goodtrailer committed:
Merge branch 'master' of https://github.com/ppy/osu into smoke

goodtrailer committed:
Return smoke key to key overlay

goodtrailer committed:
Fix test and spawn smoke immediately

peppy commented:
Is this supposed to look correct? For me it's a solid non-antialiased line which looks nothing like stable:

![2022-09-20 14 18 33@2x](https://user-images.githubusercontent.com/191335/191173744-f958d45c-54c9-4107-b965-411f911fa988.png)

goodtrailer commented:
Oh shoot I forgot to mention: for whatever reason, classic skin doesn't seem to have a smoke texture in lazer. It ends up falling back to `DefaultSmoke`. Any other legacy skin uses `LegacySmoke` though, and should look the same.

ItsShamed commented:
> This is an [osu-resources](https://github.com/ppy/osu-resources/tree/master/osu.Game.Resources/Skins/Legacy) issue though.

You can do a PR on [ppy/osu-resources](https://github.com/ppy/osu-resources)

goodtrailer commented:
> > This is an [osu-resources](https://github.com/ppy/osu-resources/tree/master/osu.Game.Resources/Skins/Legacy) issue though.
> 
> You can do a PR on [ppy/osu-resources](https://github.com/ppy/osu-resources)

Unfortunately, I do not have the original smoke texture (to my knowledge it's not public). AFAIK this is a pretty easy fix, just drag+drop the smoke texture, whoever has access to it.

ItsShamed commented:
https://osu.ppy.sh/community/forums/topics/129191?n=115

smoogipoo commented on osu.Game.Rulesets.Osu/Skinning/Smoke.cs:
```diff
@@ -0,0 +1,445 @@
+                if (Points.Count == 0)
+                    return;
+
+                QuadBatch ??= renderer.CreateQuadBatch<TexturedVertex2D>(7200, 10);
```
Probably need to cap smoke to the size of this VBO or possibly even less, since you can lag the game by drawing with no limit.

Technically this is a prime case for instanced rendering which can improve things in the future, but things are in a bit of a state of flux at the moment.

peppy committed:
Update resources

goodtrailer committed:
Cap smoke on point count + omit invisible vertices

smoogipoo committed:
Perform matrix mults on the GPU

smoogipoo committed:
Reduce the number of points

smoogipoo committed:
Dispose quad batch

smoogipoo committed:
Use British-English

smoogipoo committed:
Privatise setters

smoogipoo committed:
Expose less stuff, clean up implementation

smoogipoo committed:
Remove another invalidate

smoogipoo committed:
And another invalidate

smoogipoo commented on osu.Game.Rulesets.Osu/Skinning/Smoke.cs:
```diff
@@ -0,0 +1,384 @@
+
+        private const int max_point_count = 18_000;
+
+        public override float Height
+        {
+            get => base.Height = BottomRight.Y - TopLeft.Y;
+            set => throw new InvalidOperationException($"Cannot manually set {nameof(Height)} of {nameof(Smoke)}.");
+        }
+
+        public override float Width
+        {
+            get => base.Width = BottomRight.X - TopLeft.X;
+            set => throw new InvalidOperationException($"Cannot manually set {nameof(Width)} of {nameof(Smoke)}.");
+        }
+
+        public override Vector2 Size
+        {
+            get => base.Size = BottomRight - TopLeft;
+            set => throw new InvalidOperationException($"Cannot manually set {nameof(Size)} of {nameof(Smoke)}.");
+        }
```
Am I understanding right that this is supposed to make the smoke drawable only take up the amount of space of the smoke itself?

If so, I find this unnecessary - make the smoke take up the entire screen. Removes quite a bit of code here.

goodtrailer replied:
Yep, you're correct. I did this because a) paths/sliders do this as well, and b) it's a bit annoying to use the visual inspector (ctrl f1) when drawable bounds don't visually fit.

smoogipoo commented on osu.Game.Rulesets.Osu/Skinning/Smoke.cs:
```diff
@@ -0,0 +1,384 @@
+            }
+        }
+
+        protected abstract double LifetimeAfterSmokeEnd { get; }
```
This one needs documentation, the name isn't enough to tell me what this is supposed to represent.

smoogipoo commented on osu.Game.Rulesets.Osu/Skinning/Smoke.cs:
```diff
@@ -0,0 +1,384 @@
+        private float totalDistance;
+        private Vector2? lastPosition;
+
+        private const int max_point_count = 18_000;
```
I've reduced this from 72000 because that's a number of quads that's reached well after my FPS dips below 30.

smoogipoo reviewed (changes requested):
I've done a bit of cleanup, mostly focused around the drawnode / invalidation logic.

goodtrailer commented on osu.Game.Rulesets.Osu/Skinning/Default/DefaultSmoke.cs:
```diff
@@ -0,0 +1,65 @@
+
+namespace osu.Game.Rulesets.Osu.Skinning.Default
+{
+    public class DefaultSmoke : Smoke
```
Bad visually? I separated default/legacy and gave default temporary effects assuming that default skin (or I guess triangles/argon?) would have special smoke effects, since they have flashy effects that are impossible to achieve with legacy skins.

smoogipoo replied:
Bad visually, yeah... Maybe replace the default implementation with it and use that everywhere? Is the only suggestion I have for now. Curious what @peppy thinks of it.

goodtrailer committed:
Replace LifetimeAfterSmokeEnd with abstract LifetimeEnd

goodtrailer committed:
Lock smoke bounds to playfield

ItsShamed commented:
The re-fade animation looks a bit chunky in this PR. I don't know if this is something which can be fixed looking at how this is implemented. Not a big deal though, just something that might surprise players that are particularly familiar with stable's smoke behaviour.

| Stable | This PR |
| -- | -- |
| ![stable](https://user-images.githubusercontent.com/39100084/191466085-60884418-b88a-4d29-b837-693d44b0422a.gif) | ![lazer](https://user-images.githubusercontent.com/39100084/191466155-2ef44588-3170-4a64-9600-abbe5dcf931a.gif) |

goodtrailer commented:
Oops, forgot to mention this as well. The refade animation works by essentially replaying the smoke at 3x speed. As I understand it, stable ignores time and just refades one particle every 2ms. I found stable's effect slightly unintuitive and changed it (which also made implementation slightly easier). I don't think it'll be hard to change back to stable's behavior though, if wanted.

peppy committed:
Merge branch 'master' into smoke

peppy commented:
Is there no test setup for this? Could you please add one?

peppy commented:
The smoke trails are never expired. They stay in the draw hierarchy forever.

As proposed, please make the default use the legacy implementation for now. The default barebones seems non-useful as it doesn't look good. If we're going to be make a new default smoke, the legacy implementation would be a starting point.

peppy reviewed (changes requested):
As mentioned

goodtrailer committed:
Fix `LegacySmoke` alpha calculations

goodtrailer committed:
Remove skinnable parents at the same time as their smoke children

goodtrailer committed:
Remove unnecessary `IsActive` variable

goodtrailer commented:
> As proposed, please make the default use the legacy implementation for now. The default barebones seems non-useful as it doesn't look good. If we're going to be make a new default smoke, the legacy implementation would be a starting point.

What's the best approach to this? One is to just have `DefaultSmoke` inherit `LegacySmoke` for now, which is easy. The other is to move `LegacySmoke`'s implementation to `Smoke`, but then a lot of implementation details get carried over which only make sense if derivatives (`DefaultSmoke`/`ArgonSmoke`) are going to have basically the same effects, just different timings.

goodtrailer committed:
Add barebones test for smoke

goodtrailer committed:
Fix InspectCode issues

peppy commented:
I think it's fine to move everything to the base implementation until we decide to implement in a different way. It was tuned in stable to feel good, and already does feel pretty good. Making something that feels better (aka for argon or otherwise) will take some more thinking and effort. We can split out the code at that point.

goodtrailer committed:
Move `LegacySmoke` animation implementation to `Smoke`

peppy commented on osu.Game.Rulesets.Osu/UI/SmokeContainer.cs:
```diff
@@ -29,6 +31,8 @@ public bool OnPressed(KeyBindingPressEvent<OsuAction> e)
         {
             if (e.Action == OsuAction.Smoke)
             {
+                Logger.Log("holy moly");
```
?

peppy commented on WeatherConfig.dat:
```diff
@@ -0,0 +1,3 @@
+5363943
```
You may want to check your commits before pushing.

goodtrailer replied:
yikes

goodtrailer committed:
Remove things that should've been removed before push

peppy committed:
Update default cursor smoke implementation to use a texture

peppy committed:
Rename classes to better emphasise that `Smoke` is a single trail segment

peppy committed:
Refactor to avoid DI / event flow

There's always one active smoke segment and it's the direct child of
`SmokeContainer`. This can be simplified as such.

peppy committed:
Simplify `SmokeContainer` lifetime logic

peppy committed:
Set `LifetimeEnd` once rather than computing on every access

peppy committed:
Remove custom radius implementation

peppy committed:
Privatise some fields

peppy committed:
Update resources

peppy committed:
Merge branch 'master' into smoke

goodtrailer committed:
Add more smoke tests

goodtrailer committed:
Rename `OsuSkinComponents.SmokeTrail` to `CursorSmoke`

goodtrailer committed:
Make smoke skinnable lifetime more robust

goodtrailer commented:
Seems good.

Noticed that during replays, rewinding to before a smoke was spawned doesn't actually destroy it nor its parent skinnable, since `LifetimeStart` is never set. Realized that instead of all the extra `LifetimeStart` and `LifetimeEnd` logic, it would probably be easier to just derive `SkinnableDrawable` and override those properties to point to the child's, i.e. the `SmokeSegment`'s. At least, it seems more robust/simple to me, and it does fix the rewind issue.

(Also did a bit of renaming and added some a couple more tests.)

peppy commented:
One thing to note is that with the segments expiring, rewinding may miss some segments due to the way this is implemented. I'm not sure how easy this would be to fix, but maybe something as a follow-up effort.

https://user-images.githubusercontent.com/191335/194204340-3e43986b-6778-40e7-ad5d-8f09508afe31.mp4

peppy commented:
Going to get this in as it is. We can iterate towards fixing the remaining issues post-fact.

Joehuu commented:
"addresses" is not a keyword to close. We usually use that for discussions as you can't close those. Some keywords to close are "closes", "resolves", or "fixes".

darkralts commented:
YEEEES YEWSSKFOKSLADJKFSD'KA OIIA MOPFM, E[IOJKF MPOEWAMK FOWAO WHYESSS THE SOME TRAISLS!! THEY ARE FINALLY HERE!! THANK YOU PEPPY!

darkralts commented:
IVE BEEN WAITING MY ENTIRE LIFE FOR THIS MOMENT
